src/wallet/wallet_tx_builder.cpp

```diff
@@ -177,25 +203,248 @@ SpendableInputs WalletTxBuilder::FindAllSpendableInputs(
+    if (spendable.HasTransparentCoinbase()) {
+        return tl::make_unexpected(ChangeNotAllowedError(spendable.Total(), targetAmount));
+    }
+
+    examine(changeAddr, match {
+        // TODO: Once we can add Sprout change to `resolvedPayments`, we don’t need to pass
+        //       `changeAddr` around the rest of these functions.
+        [](const libzcash::SproutPaymentAddress&) {},
+        [](const libzcash::SproutViewingKey&) {},
+        [&](const auto& sendTo) {
+            resolvedPayments.AddPayment(
+                    ResolvedPayment(std::nullopt, sendTo, changeAmount, std::nullopt, true));
+        }
+    });
+
+    return {};
+}
+
+/// On the initial call, we haven’t yet selected inputs, so we assume the outputs dominate the
+/// actions.
+///
+/// 1. calc fee using only resolvedPayments to set a lower bound on the actual fee
+///    • this also needs to know which pool change is going to, so it can determine what the fee is
+///      with change _if_ there is change
+/// 2. iterate over LimitToAmount until the updated fee (now including spends) matches the expected
+///    fee
+tl::expected<
+    std::tuple<SpendableInputs, CAmount, std::optional<ChangeAddress>>,
+    InputSelectionError>
+WalletTxBuilder::IterateLimit(
+        CWallet& wallet,
+        const ZTXOSelector& selector,
+        const TransactionStrategy strategy,
+        CAmount sendAmount,
+        CAmount dustThreshold,
+        const SpendableInputs& spendable,
+        Payments& resolved,
+        bool afterNU5) const
+{
+    SpendableInputs spendableMut;
+
+    auto previousFee = MINIMUM_FEE;
+    auto updatedFee = CalcZIP317Fee(std::nullopt, resolved.GetResolvedPayments(), std::nullopt);
+    // This is used to increase the target amount just enough (generally by 0 or 1) to force
+    // selection of additional notes.
+    CAmount bumpTargetAmount{0};
+    std::optional<ChangeAddress> changeAddr;
+    CAmount changeAmount{0};
+    CAmount targetAmount{0};
+
+    do {
+        // NB: This makes a fresh copy so that we start from the full set of notes when we re-limit.
+        spendableMut = spendable;
+
+        targetAmount = sendAmount + updatedFee;
+
+        // TODO: the set of recipient pools is not quite sufficient information here; we should
+        // probably perform note selection at the same time as we're performing resolved payment
+        // construction above.
+        bool foundSufficientFunds =
+            spendableMut.LimitToAmount(
+                    targetAmount + bumpTargetAmount,
+                    dustThreshold,
+                    resolved.GetRecipientPools());
+        changeAmount = spendableMut.Total() - targetAmount;
+        if (foundSufficientFunds) {
+            // Don’t want to generate a change address if we don’t need one (because it could be
+            // fresh) and once we generate it, hold onto it. But we still don’t have a guarantee
+            // that we won’t end up discarding it.
```

Comment:
```suggestion
            // Don't want to generate a change address if we don’t need one (because it could be
            // fresh) and once we generate it, hold onto it. But we still don't have a guarantee
            // that we won't end up discarding it.
```
Non-blocking.